[4.0] Add the possibility to have a custom login logo

coolcat-creations opened the pull request:
This PR Adds an Option in the Backend Template to add a custom Login Logo and to show the Sitename or not.

Sitename + Login Logo

![grafik](https://user-images.githubusercontent.com/828371/45357466-9bfcfb00-b5c6-11e8-9b9c-624643fd85b3.png)

Only Login Logo:
![grafik](https://user-images.githubusercontent.com/828371/45357509-bcc55080-b5c6-11e8-9dce-a1c921ee1c87.png)

None of both:
![grafik](https://user-images.githubusercontent.com/828371/45357542-d49cd480-b5c6-11e8-9ece-cfb54c976496.png)

Test Instructions:
1) Apply the patch
2) Change your desired settings Logo / Sitename

![grafik](https://user-images.githubusercontent.com/828371/45357587-f302d000-b5c6-11e8-8a64-748c11298adc.png)

3) logout of the backend
4) Login page should show your desired logo and or Sitename
5) Try with different combinations

coolcat-creations committed:
Add the possibility to have a custom login logo

coolcat-creations committed:
Remove description

farrelljsec commented:
I have tested this item :white_check_mark: successfully on 3280db029531c0b49ff4ec685c9f38278c8b0ad1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22084">issues.joomla.org/tracker/joomla-cms/22084</a>.</sub>

brianteeman commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,16 @@
+						label="TPL_ATUM_LOGINLOGO_LABEL" />
+			</fieldset>
+		</fields>
+	</config>
```
please follow the code style
https://developer.joomla.org/coding-standards/xml.html

brianteeman commented:
The accessibility team have requested that this page has an H1 title - this PR removes that
also any image must have an alt description OR if it is purely decorative then it must have `alt=""`

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on d53bda4d5e9ca4ea94e97b10dfeceb84d318ff73<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22084">issues.joomla.org/tracker/joomla-cms/22084</a>.</sub>

coolcat-creations committed:
Change replacing h1 to adding an image if exists

coolcat-creations commented:
Changed to adding the custom logo if exists

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -73,8 +77,11 @@
+					<div class="login-box-header text-center">
+						<h1 class="m-0"><?php echo $sitename; ?></h1>
+						<?php if ($loginLogo) : ?>
+						<img src="<?php echo JURI::root() . '/' . $loginLogo; ?>" class="img-fluid my-2" alt="<?php echo $sitename; ?>">
```
is $sitename really an appropriate description of the image?

coolcat-creations replied:
What do you suggest?

brianteeman replied:
You either have to add a field to the xml (which still needs fixing) for the alt description or

>  if it is purely decorative then it must have alt=""

coolcat-creations committed:
Correct Code Style

coolcat-creations committed:
Remove alt description

coolcat-creations commented:
Hm I copied a field from isis template, seems no code style is followed there, thats where the mistake came from.

brianteeman commented:
dont see any issue in isis or i would fix it

```
				<field 
					name="loginLogoFile" 
					type="media" 
					label="TPL_ISIS_LOGIN_LOGO_LABEL"
					description="TPL_ISIS_LOGIN_LOGO_DESC"
					class="" 
					default=""
				 />
```
https://github.com/joomla/joomla-cms/blob/eb68e01fada0ceac8695c386c9f9ac143db2f079/administrator/templates/isis/templateDetails.xml#L109

crommie commented:
I have tested this item :white_check_mark: successfully on ba2709d9369d5784669ae80f8829f255f0ebd1b6<br /><br />Followed instructions. Result as expected. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22084">issues.joomla.org/tracker/joomla-cms/22084</a>.</sub>

coolcat-creations commented:
@brianteeman 
<img width="983" alt="grafik" src="https://user-images.githubusercontent.com/828371/45255666-2b699a80-b38a-11e8-8ecf-d7567a458540.png">

brianteeman commented:
no idea what you have on your computer but it is not the same code we have in github
https://github.com/joomla/joomla-cms/blob/eb68e01fada0ceac8695c386c9f9ac143db2f079/administrator/templates/isis/templateDetails.xml#L51

brianteeman reviewed (approved):
approve language change

SharkyKZ commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -13,3 +13,4 @@ TPL_ATUM_TOOLBAR="Toolbar"
 TPL_ATUM_VIEW_SITE="View site"
 TPL_ATUM_XML_DESCRIPTION="Continuing the Egyptian god/goddess theme (Khepri from 1.5, Hathor from 1.6, Isis from 3.0), Atum is the Joomla 4 administrator template based on Bootstrap 4."
 TPL_BACK_TO_CONTROL_PANEL="Back to Control Panel"
+TPL_ATUM_LOGINLOGO_LABEL="Login Logo"
```
Sort keys alphabetically.

SharkyKZ commented on administrator/templates/atum/login.php:
```diff
@@ -73,8 +77,11 @@
+					<div class="login-box-header text-center">
+						<h1 class="m-0"><?php echo $sitename; ?></h1>
+						<?php if ($loginLogo) : ?>
+						<img src="<?php echo JURI::root() . '/' . $loginLogo; ?>" class="img-fluid my-2" alt="">
```
Indent. Change `JURI` to `Uri`. Remove slash afterwards.

SharkyKZ commented:
@coolcat-creations oops, I meant to say remove if specified.

I think @brianteeman means that image alt could be added as another param.

coolcat-creations commented:
@SharkyKZ as far I understood he said to keep the h1 title for a11y.

brianteeman commented:
BTW the xml you copied came from a very very old version of joomla - thats why you had the problem

tonypartridge commented:
@coolcat-creations can we add a toggle for the Page heading? Default on of course.

coolcat-creations commented:
@tonypartridge if it’s an accessibility thing I rather would not add an option to remove it. Only thing is that I do not understand why the sitename increases the accessibility- I rather see here then a title that explains where I am. Something like „Joomla Backend Login Page“ - this can be written in small letters and below / sitename or Logo or both.

tonypartridge commented:
But to me there are cases where design wins over accessibility. In terms of a company which has no employee with accessibility needs, so a customising toggle to put a logo in and hide H1 tag would be much more beneficial.   
  

  
For a11y I would have thought it would need an Administrator string attached so they know they are in the backend?   
  

  

  

  
  

  
  
>   
> On Sep 9, 2018 at 6:33 pm,  <Elisa Foltyn (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> @tonypartridge (https://github.com/tonypartridge)  if it’s an accessibility thing I rather would not add an option to remove it. Only thing is that I do not understand why the sitename increases the accessibility- I rather see here then a title that explains where I am. Something like „Joomla Backend Login Page“ - this can be written in small letters and below / sitename or Logo or both.
>
>   
>
> —
>  You are receiving this because you were mentioned.
>  Reply to this email directly,   view it on GitHub (https://github.com/joomla/joomla-cms/pull/22084#issuecomment-419731637), or   mute the thread (https://github.com/notifications/unsubscribe-auth/ABVglkDweYD_rT3tmVRKnpunj8BfuNzBks5uZVD9gaJpZM4Wf5Nw).
>
>     
>

coolcat-creations commented:
Yes I have the similar thought about that it makes more sense to tell people where they are instead of output of the sitename. Doesn’t make sense to me regarding a11y. I will have a thought about how to solve your request and remain a11y compliant :)

coolcat-creations committed:
Add option to show Sitename or not
Add a11y h1 to show the page where the User is at
Refactor some of the code

coolcat-creations commented:
@tonypartridge I added the functionality to show the sitename or not and added the A11y h1 with the information where the User is at...

What I will have to do in another PR is to refactor the whole login page because it has a lot of useless additional classes and is not using the BS4 that is already providing those classes.

coolcat-creations commented:
Testing Instructions and PR Description is updated.

brianteeman commented:
I am not in favour of this change. If a site builder wants to customise the login page they should create an override

tonypartridge commented:
I have to disagree with your @brianteeman  this is a nice and easy way to provide some basic branding to the admin area for the clients. 

You shouldn't need complex template overrides to give a basic style to the backend login page which this does nicely.

coolcat-creations commented:
I agree to @tonypartridge 
1) Easy Layout adjustments should be able to be done with few clicks - I would even add a few more (Background Color and Image) - as well as changes for the backend colors. (like described in the backend concept - Please read the concept for that)
2) There is no possibility to override a core template. If so, then I missed possibly a huge feature. You can only copy and assign the copy and then some day you will be completely out-of-date with your copy. 
3) Override of a core template would be great, for that we would need to define where the overrides will be stored. I can try to add this feature so that we have basic options in the template and advanced options per override.

tonypartridge commented:
I have tested this item :white_check_mark: successfully on 57a1839b4f227799cb7c614e87bb14aebf1b5c0a<br /><br />Tested working as desired.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22084">issues.joomla.org/tracker/joomla-cms/22084</a>.</sub>

coolcat-creations commented:
@SharkyKZ can you retest? If this is merged I will work on the other mentioned points in seperate PRs

tonypartridge commented:
@coolcat-creations just to add something else to your list... we could do with enhancing the note field to be 100% width, the label jumps onto another line at present. Maybe an if only label and no description fill 100% like if we set hr="true"

franz-wohlkoenig commented:
Ready to Commit after two successful tests.

coolcat-creations commented:
@tony and @crommie thank you for testing and the suggestions.
I will create another PR with some general code improvements on the login view. This is independant of this PR and I will do it as soon as this is merged.

mbabker commented:
> Override of a core template would be great, for that we would need to define where the overrides will be stored. I can try to add this feature so that we have basic options in the template and advanced options per override.

No.  Just, no.  The template is the primary source of truth for all things markup.  If we start introducing systems where template markup can be overridden, it will make things even more complicated than they are.

> You shouldn't need complex template overrides to give a basic style to the backend login page which this does nicely.

What defines "basic" options and what defines options that are best suited as a new template?  By your logic, almost everything that is controlled with some form of CSS variable (default font size, font family, color schemes of every element, etc.) could be considered "basic" and next thing you know it the template style edit page for the backend template has about 300 options.

> You can only copy and assign the copy and then some day you will be completely out-of-date with your copy.

If you are copying a template, you are in essence forking it.  Why does it matter if the source template continues to evolve after you've copied the template?  At the point you copied the template it is your's to customize however you please, which includes the maintenance work of synchronizing changes from the source template if you so desire.

SharkyKZ commented:
@coolcat-creations code should be cleaned before merging.

mbabker commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,35 @@
+		<fields name="params">
+			<fieldset name="advanced">
+
+				<field
```
This field should be the fieldset's description or something, this is major misuse of fields to create a heading/description for a fieldset/tab.

mbabker commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,35 @@
+				/>
+
+				<field
+						name="showSitename"
```
Codestyle - One indent for each line (there are two here and in follow on lines)

mbabker commented on administrator/templates/atum/login.php:
```diff
@@ -41,6 +41,11 @@
+$showSitename = $this->params->get('showSitename', '1');
+$loginLogo = $this->params->get('loginLogo', '');
+
+
```
Codestyle - Too many blank lines

coolcat-creations commented:
>>     Override of a core template would be great, for that we would need to define where the overrides will be stored. I can try to add this feature so that we have basic options in the template and advanced options per override.

> No. Just, no. The template is the primary source of truth for all things markup. If we start introducing systems where template markup can be overridden, it will make things even more complicated than they are.

I am ok with your no. I just was trying to fulful something @brianteeman suggested as solution.

mbabker commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -4,12 +4,17 @@
 
 ATUM="Atum Administrator template"
+TPL_ATUM_ADMIN_USER_MENU="User Menu"
+TPL_ATUM_BACKEND_LOGIN="Joomla Administrator Loginpage"
```
"Loginpage" is not a word.

coolcat-creations commented:
>> You shouldn't need complex template overrides to give a basic style to the backend login page which this does nicely.

> What defines "basic" options and what defines options that are best suited as a new template? By your logic, almost everything that is controlled with some form of CSS variable (default font size, font family, color schemes of every element, etc.) could be considered "basic" and next thing you know it the template style edit page for the backend template has about 300 options.

Basic is defined by how you can adjust the backend to a customer or site integrator. This was defined and agreed by the marketing department to increase the possibilities to add the own branding into the template. Basic Options IMHO are: Logo and Colorscheme.

coolcat-creations commented:
>>    You can only copy and assign the copy and then some day you will be completely out-of-date with your copy.

> If you are copying a template, you are in essence forking it. Why does it matter if the source template continues to evolve after you've copied the template? At the point you copied the template it is your's to customize however you please, which includes the maintenance work of synchronizing changes from the source template if you so desire.

Thats the reason why a core template should have some basic options. Why should I fork the template just because I want to change the displayed logo. Does not make any sense and is an overkill in managing that.

coolcat-creations commented:
> @coolcat-creations code should be cleaned before merging.

The code cleanup will be done in a seperate PR because the whole classes of the login page have to be rewritten. There is so much unneeded additional classes there...

The scope of the PR was to add the options as successfuly tested, and I will create a new PR so change the markup afterwards.

mbabker commented:
"Basic Options" is still an arbitrary measure, and that's kind of my point.  It's not that the options are bad necessarily, but different viewpoints are going to suggest that different customizations should be offered through the template style edit page and not require custom CSS or whatnot.  If there is an agreed upon list of changes that will be supported through the template style edit page, that's fine, I just don't want us entertaining options for too many elements personally.

coolcat-creations committed:
Changes as requested by @mbabker review

coolcat-creations commented:
@mbabker 

Login Page: Vendor Logo, Show/hide sitename, background color, background image
Backend: Setup colors

For more changes:
Maybe a possibility to add custom.scss

For (even) more changes:
Admin has to create a fork...

Ok?

coolcat-creations committed:
Own Tab for Login Page Settings as suggested by @mbabker

coolcat-creations commented:
@mbabker thank you - done - I planned to have all settings on one tab in future thats why I used the spacer, but changed it now to own tab

brianteeman reviewed (commented):
Please do not merge yet. The language strings are not correct. I won't get a chance to review until tomorrow.

mbabker commented:
(Marking my requested changes as completed, not that PR is 100% "good")

Quy commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,29 @@
+					class="btn-group btn-group-yesno"
+					default="1"
+					filter="integer"
+				>
```
Indent

coolcat-creations replied:
@Quy you mean that the whole config block has to have an indent generally?

SharkyKZ replied:
Just line 58. `>` has to align with `<` of `<option>` on the next line.

coolcat-creations replied:
Thank you

franz-wohlkoenig commented:
> The language strings are not correct.

@brianteeman should RTC set back on Pending?

coolcat-creations committed:
Indent

brianteeman commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -4,12 +4,17 @@
 
 ATUM="Atum Administrator template"
+TPL_ATUM_ADMIN_USER_MENU="User Menu"
+TPL_ATUM_BACKEND_LOGIN="Joomla Administrator login"
```
Please capitalise all words as it is a label

brianteeman commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -4,12 +4,17 @@
 TPL_ATUM_INSTALLER="Installer"
 TPL_ATUM_ISFREESOFTWARE="Joomla is free software released under the GNU General Public License."
+TPL_ATUM_LOGIN_PAGE_SETTINGS_LABEL="Login Page Settings"
+TPL_ATUM_LOGINLOGO_LABEL="Login Logo"
```
For readability can you change this to LOGIN_LOGO and then re-alphasort

brianteeman commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -4,12 +4,17 @@
+TPL_ATUM_LOGINLOGO_LABEL="Login Logo"
 TPL_ATUM_MENU="MENU"
-TPL_ATUM_ADMIN_USER_MENU="User Menu"
+TPL_ATUM_SHOWSITENAME_LABEL="Show site name"
```
Please capitalise all words as it is a label
Shouldnt the label key be simply TPL_ATUM_SITENAME_LABEL

brianteeman replied:
And as we no longer use words like show in the label please just make it "Site Name"

brianteeman commented on administrator/templates/atum/language/en-GB/en-GB.tpl_atum.ini:
```diff
@@ -4,12 +4,17 @@
 TPL_ATUM_MENU="MENU"
-TPL_ATUM_ADMIN_USER_MENU="User Menu"
+TPL_ATUM_SHOWSITENAME_LABEL="Show site name"
+TPL_BACK_TO_CONTROL_PANEL="Back to Control Panel"
```
Please put this line back where it was so that the strings are alpha sorted

coolcat-creations replied:
I dont' understand, I put it at the top because it starts with A... where do you want me to put it?

SharkyKZ replied:
`TPL_BACK_TO_CONTROL_PANEL` should be at the bottom.

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -72,11 +76,24 @@
 				<div id="element-box" class="login-box">
-					<div class="login-box-header">
-						<h1 class="text-center m-0"><?php echo $sitename; ?></h1>
+
```
remove blank line

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -72,11 +76,24 @@
-						<h1 class="text-center m-0"><?php echo $sitename; ?></h1>
+
+					<?php if ($showSitename || $loginLogo) : ?>
+
```
remove blank line

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -72,11 +76,24 @@
+						<?php endif; ?>
 					</div>
-					<div class="login-box-body">
+
```
remove blank line

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -72,11 +76,24 @@
-					<div class="login-box-body">
+
+					<?php endif; ?>
+
```
remove blank line

brianteeman commented on administrator/templates/atum/login.php:
```diff
@@ -72,11 +76,24 @@
+
+					<?php endif; ?>
+
+
```
remove blank line

brianteeman commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,29 @@
+					name="showSitename"
+					type="radio"
+					label="TPL_ATUM_SHOWSITENAME_LABEL"
+					class="btn-group btn-group-yesno"
```
In J4 this should be 
`			class="switcher"
`

brianteeman commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,29 @@
+					default="1"
+					filter="integer"
+					>
+					<option value="1">JYES</option>
```
As we no longer use words like Show in the label this should be JSHOW and JHIDE

brianteeman commented:
Please set back to pending

coolcat-creations committed:
Applied codereviews

SharkyKZ commented:
Last commit broke codestyle in `login.php`.

coolcat-creations committed:
Fix Codestyle

brianteeman commented on administrator/templates/atum/templateDetails.xml:
```diff
@@ -44,4 +44,29 @@
+
+				<field
+					name="showSitename"
+					type="switcher"
```
You must have misunderstood my comment here https://github.com/joomla/joomla-cms/pull/22084#discussion_r216925365 and then not tested it

The type is still radio but the class should be switcher

coolcat-creations committed:
fix language string
fix wrong type

wilsonge committed:
Codestyle

wilsonge committed:
Codestyle v2

wilsonge committed:
Merge branch '4.0-dev' into add-login-logo-config-option

wilsonge commented:
Thanks guys!
